lib/matplotlib/axes/_base.py

```diff
@@ -1565,23 +1587,24 @@ def apply_aspect(self, position=None):
         xm = 0
         ym = 0
 
-        changex = (self in self._shared_y_axes and
-                   self not in self._shared_x_axes)
-        changey = (self in self._shared_x_axes and
-                   self not in self._shared_y_axes)
-        if changex and changey:
-            warnings.warn("adjustable='datalim' cannot work with shared "
-                          "x and y axes")
-            return
-        if changex:
+        shared_x = self in self._shared_x_axes
+        shared_y = self in self._shared_y_axes
+        # Not sure whether we need this check:
+        if shared_x and shared_y:
+            raise RuntimeError("adjustable='datalim' is not allowed when both"
```

Comment:
Should it not be possible to share datalim with both axes?

Reply:
When the box dimensions are fixed, as is the case with "adjustable='datalim'", it is not possible to maintain a specified aspect ratio without being able to adjust at least one of the axes and have it stick.  You can't do that when they are both shared.

Reply:
Let see if I get this straight. When adjustable is "box" they share both the datalim and the physical dim of the axis, but when adjustable is datalim it just it is the xlim or ylim that are shared?

What is the default adjustable when axes are created?

Reply:
Sharing applies only to datalim (in this context, but it is actually done via a reference to the same Axis, so locators and formatters as well as datalim are locked together).  Twinning locks the box to be identical between the twins.  The default adjustable is now box.

Reply:
Thus datalim/box are just used by the set_aspect. 

With the example @stonebig and @mwaskom posted at #10585 twinning would become a new share group. That is the diagonal plots share y-axis. If you move up in (i,i) plot then i row will also move, but if you move up in (i,j) j not i then only that row except (i,i) will move.

This of course make sharing a bit harder with the new API (if it gets accepted), since sharing is transitive.
The only way to get the same result would be to listen to LocationEvent and break/create share
as one move between the axes. Or you need to open the make_twin API to the public.

PS. @efiring I fixed such that major and minor and copy over when share is broken in the new API.